### ISSUE TITLE
Custom STEP Model Generation Error

### ISSUE THREAD

**Issue by ddkkgit** [SadilKhan/Text2CAD#31]
I would like to express my profound gratitude to the author for their significant contributions. Currently, I am endeavoring to utilize your work for the generation of custom three-dimensional models (single-step). The two primary files I am executing are test_user_input.py and json2step.py; the former is utilized to generate the sequences required for custom 3D models, while the latter is used to produce STEP files. However, during this process, even when I execute the example prompt you provided, "A rectangular prism with a hole in the middle," the results I obtain are not satisfactory, as the generated models typically manifest as rectangular prisms without any additional features.

I would like to inquire whether there might be any errors in the steps I am executing, and if so, how I might correctly generate custom STEP 3D models?

Additionally, as a supplementary note, during the debugging process, I altered the text_embedder:bert_large_uncased to load locally. While debugging, I encountered an issue with matrix size discrepancy, where a size of 1024 was required, but the actual output was 768. This issue arose in the adaptive_layer, specifically within the TextAdaptiveLayer class. Its preceding layer is text2cad.py, specifically at the code: ZE, key_padding_mask = self.base_text_embedder.get_embedding(texts). To facilitate the program's execution, I added a linear layer in adaptive_layer.py to map the feature dimensions from 768 to 1024, allowing the entire program to run. I am uncertain whether my modifications might have caused errors in the generation of CAD model sequences.

Your assistance is greatly appreciated!

**Comment by SadilKhan** [comment]
Hello,

Did you train the new Linear Layer that you added?

**Comment by ddkkgit** [comment]
I am very sorry, but I have not trained any new parameters; I directly used the pre-trained weights and vectors you made publicly available. Due to limitations in computational resources, I am afraid I do not have the capability to perform retraining. So, how can I run this project without adding a linear layer?

**Comment by SadilKhan** [comment]
> Additionally, as a supplementary note, during the debugging process, I altered the text_embedder:bert_large_uncased to load locally. While debugging, I encountered an issue with matrix size discrepancy, where a size of 1024 was required, but the actual output was 768. This issue arose in the adaptive_layer, specifically within the TextAdaptiveLayer class. Its preceding layer is text2cad.py, specifically at the code: ZE, key_padding_mask = self.base_text_embedder.get_embedding(texts). To facilitate the program's execution, I added a linear layer in adaptive_layer.py to map the feature dimensions from 768 to 1024, allowing the entire program to run. I am uncertain whether my modifications might have caused errors in the generation of CAD model sequences.
> 


Can you please upload the modified code here?

**Comment by ddkkgit** [comment]
Of course.
adaptive_layer.py
class TextAdaptiveLayer(nn.Module):
    """
    Adaptive Layer for Text Embeddings

    """

    def __init__(self, in_dim: int, out_dim: int, num_heads: int, dropout: float):
        super(TextAdaptiveLayer, self).__init__()

        # Multi-Head Self Attention for CAD Sequence
        # TODO: Check if Flash Attention is implemented
        ......
        # LayerNormalization
        self.norm_seq = nn.ModuleDict(
            {"norm_1": nn.LayerNorm(in_dim), "norm_2": nn.LayerNorm(in_dim)}
        )

        # Dropout
        self.dp_seq = nn.ModuleDict(
            {"dropout_1": nn.Dropout(dropout), "dropout_2": nn.Dropout(dropout)}
        )
        """
            Add a linear layer:self.changedim
        """
        self.changedim=nn.Linear(768,1024)

        # Feed forward Networks
        self.ffl_seq = FeedForwardLayer(input_dim=in_dim)

    def forward(
        self,
        T: Optional[torch.Tensor],
        mask_prompt_dict: dict,
        metadata: bool = False,
    ):
        """
        T: tensor of shape (bs, num_seq, emb_dim). Text Embedding
        mask_prompt_dict: dictionary with keys "attn_mask", "key_padding_mask"
        metadata: boolean. To save attention weights
        """

        # self_attn_mask_dict = mask_cad_dict.copy()

        # ? <----------  TEXT EMBEDDING SELF-ATTENTION  ---------->
        """
            Add a linear layer: self.changedim ;  Attention!
	Check
        """
        print("*******************************ATTENTION*********************************")
        # print("T.shape: ",T.shape)
        T = self.changedim(T)
        print("*******************************ATTENTION*********************************")

        T2 = self.norm_seq["norm_1"](T)  # (bs,num_seq,emb_dim)
        # exit()
I only added a linear layer named self.changedim in the above code. The specific files, text2cad.py and adaptive_layer.py, can be found at [https://github.com/ddkkgit/ERROR-ABOUT-TEXT2CAD ](https://github.com/ddkkgit/ERROR-ABOUT-TEXT2CAD).
Thank you very much for your patience in addressing the issues I raised.

**Comment by SadilKhan** [comment]
I see then it's probably the model's performance. Our current model is a bit prompt-sensitive since we have trained a single-stage text-to-3d generation architecture using very limited data (This is very uncommon in text-to-3d domain). We are working to improve the performance and robustness with newer architectures and we will soon release soon.
